MoreLinq.Test/PartitionTest.cs

```diff
@@ -0,0 +1,194 @@
+        }
+
+        [Test]
+        public void PartitionBooleanGrouping()
+        {
+            var (evens, odds) =
+                Enumerable.Range(0, 10)
+                          .GroupBy(x => x % 2 == 0)
+                          .Partition(ValueTuple.Create);
+
+            Assert.That(evens, Is.EquivalentTo(new[] { 0, 2, 4, 6, 8 }));
+            Assert.That(odds,  Is.EquivalentTo(new[] { 1, 3, 5, 7, 9 }));
+        }
+
+        [Test]
+        public void PartitionNullableBooleanGrouping()
+        {
+            var xs = new int?[] { 1, 2, 3, null, 5, 6, 7, null, 9, 10 };
+
+            var (lt5, gte5, nils) =
+                xs.GroupBy(x => x != null ? x < 5 : (bool?) null)
+                  .Partition(ValueTuple.Create);
+
+            Assert.That(lt5,  Is.EquivalentTo(new[] { 1, 2, 3 }));
+            Assert.That(gte5, Is.EquivalentTo(new[] { 5, 6, 7, 9, 10 }));
+            Assert.That(nils, Is.EquivalentTo(new int?[] { null, null }));
+        }
+
+        [Test]
+        public void PartitionBooleanGroupingWithSingleKey()
+        {
+            var (m3, etc) =
+                Enumerable.Range(0, 10)
+                          .GroupBy(x => x % 3)
+                          .Partition(0, ValueTuple.Create);
+
+            Assert.That(m3, Is.EquivalentTo(new[] { 0, 3, 6, 9 }));
+
+            using (var r = etc.Read())
+            {
+                var r1 = r.Read();
+                Assert.That(r1.Key, Is.EqualTo(1));
+                Assert.That(r1, Is.EquivalentTo(new[] { 1, 4, 7 }));
+
+                var r2 = r.Read();
+                Assert.That(r2.Key, Is.EqualTo(2));
+                Assert.That(r2, Is.EquivalentTo(new[] { 2, 5, 8 }));
+
+                r.ReadEnd();
+            }
+        }
+
+        [Test]
+        public void PartitionBooleanGroupingWitTwoKeys()
+        {
+            var (ms, r1, etc) =
+                Enumerable.Range(0, 10)
+                          .GroupBy(x => x % 3)
+                          .Partition(0, 1, ValueTuple.Create);
+
+            Assert.That(ms, Is.EquivalentTo(new[] { 0, 3, 6, 9 }));
+            Assert.That(r1, Is.EquivalentTo(new[] { 1, 4, 7 }));
+
+            using (var r = etc.Read())
+            {
+                var r2 = r.Read();
+                Assert.That(r2.Key, Is.EqualTo(2));
+                Assert.That(r2, Is.EquivalentTo(new[] { 2, 5, 8 }));
+                r.ReadEnd();
+            }
+        }
+
+        [Test]
+        public void PartitionBooleanGroupingWitThreeKeys()
+        {
+            var (ms, r1, r2, etc) =
+                Enumerable.Range(0, 10)
+                    .GroupBy(x => x % 3)
+                    .Partition(0, 1, 2, ValueTuple.Create);
+
+            Assert.That(ms, Is.EquivalentTo(new[] { 0, 3, 6, 9 }));
+            Assert.That(r1, Is.EquivalentTo(new[] { 1, 4, 7 }));
+            Assert.That(r2, Is.EquivalentTo(new[] { 2, 5, 8 }));
+            using (var r = etc.Read())
+                r.ReadEnd();
```

Comment:
I would find it clearer to `Assert.That(etc, Is.Empty)`, any particular reason to use the `SequenceReader` in this context?